From 0c29797bf513312e03e54c33277829576d2be7d9 Mon Sep 17 00:00:00 2001
From: Raphael Mounier <mounierr07@gmail.com>
Date: Sat, 17 Jun 2023 10:45:46 +0200
Subject: [PATCH] vold : Exfat can be mounted with "exfat" kernel fs driver

Exfat can be mounted with "exfat" kernel fs driver, or
 "sdfat" or "texfat" (Samsung and Sony variants)

Change-Id: Icae06e870aec227f051098f5d4e9a68564820bf8
---
 Utils.cpp    |  5 +++++
 fs/Exfat.cpp | 11 +++++------
 2 files changed, 10 insertions(+), 6 deletions(-)

diff --git a/Utils.cpp b/Utils.cpp
index 8ce1370..d6c10a5 100644
--- a/Utils.cpp
+++ b/Utils.cpp
@@ -1004,6 +1004,11 @@ bool IsFilesystemSupported(const std::string& fsType) {
     /* fuse filesystems */
     supported.append("fuse\tntfs\n");
 
+    /* treat sdfat/texfat as exfat */
+    if (supported.find("sdfat\n") != std::string::npos || supported.find("texfat\n") != std::string::npos) {
+        supported.append("\texfat\n");
+    }
+
     return supported.find(fsType + "\n") != std::string::npos;
 }
 
diff --git a/fs/Exfat.cpp b/fs/Exfat.cpp
index e38ca21..8be7ac2 100644
--- a/fs/Exfat.cpp
+++ b/fs/Exfat.cpp
@@ -34,11 +34,10 @@ static const char* kMkfsPath = "/system/bin/mkfs.exfat";
 static const char* kFsckPath = "/system/bin/fsck.exfat";
 
 bool IsSupported() {
-    return (access(kMkfsPath, X_OK) == 0) && 
-           (access(kFsckPath, X_OK) == 0) &&
+    return access(kMkfsPath, X_OK) == 0 && access(kFsckPath, X_OK) == 0 &&
            (IsFilesystemSupported("exfat") ||
-           IsFilesystemSupported("sdfat") ||
-           IsFilesystemSupported("texfat"));
+            IsFilesystemSupported("texfat") ||
+            IsFilesystemSupported("sdfat"));
 }
 
 status_t Check(const std::string& source) {
@@ -64,9 +63,9 @@ status_t Mount(const std::string& source, const std::string& target, int ownerUi
     auto mountData = android::base::StringPrintf("uid=%d,gid=%d,fmask=%o,dmask=%o", ownerUid,
                                                  ownerGid, permMask, permMask);
 
-    const char *fs = "exfat";
-    if (IsFilesystemSupported("texfat")) fs = "texfat";
+    const char* fs = "exfat";
     if (IsFilesystemSupported("sdfat")) fs = "sdfat";
+    if (IsFilesystemSupported("texfat")) fs = "texfat";
     if (mount(source.c_str(), target.c_str(), fs, mountFlags, mountData.c_str()) == 0) {
         return 0;
     }
-- 
2.25.1

